### PR TITLE
Add method ArrayHelper::setValue() Writes a value to the array at the specified path

### PR DESCRIPTION
**Pull request by lav45**
| Q             | A
| ------------- | ---
| Is bugfix?    | no
| New feature?  | yes
| Breaks BC?    | no
| Tests pass?   | no
| Fixed issues  | #6644

Writes a value to the array at the specified path.
If the array key specified path is missing, then it will be created and it is assigned an appropriate value.
If the key exists, then it will be overwritten.

### CODE REVIEW

File: framework/helpers/BaseArrayHelper.php [hunk 1]
```diff
@@ -215,6 +215,75 @@ public static function getValue($array, $key, $default = null)
...
+     *  [
+     *      'key' => [
+     *          'in' => [
+     *              'arr' => 'val'
+     *          ]
+     *      ]
+     *  ]
+     * ```
+     *
+     * @param array $array
```
Comment (samdark) [inline framework/helpers/BaseArrayHelper.php]: Need descriptions.

File: framework/helpers/BaseArrayHelper.php [hunk 2]
```diff
@@ -215,6 +215,75 @@ public static function getValue($array, $key, $default = null)
...
+     * @param mixed $value
+     * @return array
+     */
+    public static function setValue(array $array, $key, $value)
+    {
+        if (($pos = strpos($key, '.')) !== false) {
+            $left_key = substr($key, 0, $pos);
+            $right_key = substr($key, $pos + 1);
+
+            if (isset($array[$left_key])) {
```
Comment (samdark) [inline framework/helpers/BaseArrayHelper.php]: Variables should be named in camelCase such as `$leftKey`.

**Review by samdark — CHANGES_REQUESTED** [review]

**Review by samdark — CHANGES_REQUESTED** [review]
I think it should behave similar to ArrayHelper::getValue() i.e. support specifying path as an array.

**Comment by LAV45** [comment]
you can add support for both options

**Comment by samdark** [comment]
Please do.

**Comment by cebe** [comment]
Looks good to me in general. Have you considered passing the array by reference?
Using it like this might be more useful:

```php
$array = [ ... ];
ArrayHelper::setValue($array, 'components.db', [ ... ]);
ArrayHelper::setValue($array, 'components.cache', [ ... ]);
// ...
```

compared to:

```php
$array = [ ... ];
$array = ArrayHelper::setValue($array, 'components.db', [ ... ]);
$array = ArrayHelper::setValue($array, 'components.cache', [ ... ]);
// ...
```

or are there good reasons/use cases where you'd want to keep the original array?

**Comment by LAV45** [comment]
Check description on the `docs/guide-ru/helper-array.md` if everything is OK then I translate it in English

File: docs/guide-ru/helper-array.md [hunk 3]
```diff
@@ -47,6 +47,54 @@ $fullName = ArrayHelper::getValue($user, function ($user, $defaultValue) {
...
+        'in' => ['k' => 'value']
+    ]
+];
+
+ArrayHelper::setValue($array, 'key.in', ['arr' => 'val']);
+// путь можно по которому записано новое значение в `$array` может указать как массив  
+ArrayHelper::setValue($array, ['key', 'in'], ['arr' => 'val']);
+```
+
+В результате исходное значение `$array['key']['in']` будет перезаписано но новое
```
Comment (samdark) [inline docs/guide-ru/helper-array.md]: но новое → новым

File: docs/guide-ru/helper-array.md [hunk 4]
```diff
@@ -47,6 +47,54 @@ $fullName = ArrayHelper::getValue($user, function ($user, $defaultValue) {
...
+
+```php
+$array = [
+    'key' => [
+        'in' => ['k' => 'value']
+    ]
+];
+
+ArrayHelper::setValue($array, 'key.in', ['arr' => 'val']);
+// путь можно по которому записано новое значение в `$array` может указать как массив
```
Comment (samdark) [inline docs/guide-ru/helper-array.md]: путь можно по которому записано новое значение → путь для записи значения

**Review by samdark — CHANGES_REQUESTED** [review]
Wording should be fixed but overall info in the guide is good. Please translate into English.

File: framework/helpers/BaseArrayHelper.php [hunk 5]
```diff
@@ -215,6 +215,84 @@ public static function getValue($array, $key, $default = null)
...
+     *  the path can be described by a string when each key should be separated by a dot
+     *  you can also describe the path as an array of keys
+     *  if the path is null then `$array` will be assigned the `$value`
+     * @param mixed $value the value to be written
+     * @return array a new array with the value written
+     * @since 2.0.13
+     */
+    public static function setValue(&$array, $path, $value)
+    {
+        if (is_null($path)) {
```
Comment (cebe) [inline framework/helpers/BaseArrayHelper.php]: `is_null($path)` -> `$path === null`

File: framework/helpers/BaseArrayHelper.php [hunk 6]
```diff
@@ -215,6 +215,84 @@ public static function getValue($array, $key, $default = null)
...
+     *  ]
+     * ```
+     *
+     * @param array $array the array to write the value to
+     * @param string|array|null $path the path of where do you want to write a value to `$array`
+     *  the path can be described by a string when each key should be separated by a dot
+     *  you can also describe the path as an array of keys
+     *  if the path is null then `$array` will be assigned the `$value`
+     * @param mixed $value the value to be written
+     * @return array a new array with the value written
```
Comment (cebe) [inline framework/helpers/BaseArrayHelper.php]: there is no need to return the array if it is passed by reference.

**Review by cebe — CHANGES_REQUESTED** [review]

**Review by dynasource — APPROVED** [review]
nice work

File: docs/guide-ru/helper-array.md (hunk 4 above)
Comment (samdark) [inline docs/guide-ru/helper-array.md]: может → можно

**Review by samdark — APPROVED** [review]

**Comment by samdark** [comment]
Merging since @cebe comments were addressed.

**Comment by samdark** [comment]
Merged. Thank you!

**Comment by mattether** [comment]
Do you think this should work for objects, as `getValue` does?

I've put in a dirty fix, which works on ActiveRecords, checking whether the `$array` has an attribute for the first `$keys`

```php
//...
if (is_object($array) && $array instanceof \yii\db\ActiveRecord && $array->hasAttribute($keys[0]))
//...
```

Obviously could be a lot better, just wondering whether you think the core should handle it?

**Comment by samdark** [comment]
Not sure about it. What's use case?

**Comment by samdark** [comment]
Don't know. Let's wait till it's asked for i.e. when it's really required for a project.

**Pull request merged**